File: Dynamic-Programming/Maximum-Sum-Circular-Subarray/MaximumSumCircularSubarray.js
/**
 * @param {number[]} A
 * @return {number}
 */

var maxSubarraySumCircular = function (A) {
    let k = maxSubArrayKadane(A);
    let CS = 0;
    for (let i = 0; i < A.length; i += 1) {
        CS += A[i];
        A[i] = A[i] * -1;
    }
    CS += maxSubArrayKadane(A);
    if (CS > k && CS != 0) {
        return CS;
    } else {
        return k;
    }
};

var maxSubArrayKadane = function (A) {
    let mx = Number.MAX_SAFE_INTEGER * -1;
    let val = mx;
    for (let i = 0; i < A.length; i += 1) {
        val = Math.max(A[i], A[i] + val);
        if (val > mx) {
            mx = val;
        }
    }
    return mx;
};

File: Dynamic-Programming/Maximum-Sum-Circular-Subarray/MaximumSumCircularSubarray.py
import sys


class Solution:
    def maxSubarraySumCircular(self, A: List[int]) -> int:
        k = self.__maxSubArrayKadane(A)

        CS: int = 0
        for i in range(len(A)):
            CS += A[i]
            A[i] = A[i] * -1

        CS += self.__maxSubArrayKadane(A)

        return CS if CS > k and CS != 0 else k

    def __maxSubArrayKadane(self, nums: list):
        mx = val = sys.maxsize * -1
        for i in nums:
            val = max(i, i + val)
            if val > mx:
                mx = val
        return mx